Add ability to specify specific line or line blocks

ostollmann opened the pull request:
Close #3 

This adds the possibility to specify specific lines to be imported:

````
```js file=./__fixtures__/say-hi.js#L2
```
````

````
```js file=./__fixtures__/say-hi.js#L1-L2
```
````

Note: couldn't get yarn to work, but tests passed using npm.

ostollmann committed:
Add support for specifying lines to be extracted

ostollmann committed:
added test using just single line number

kevin940726 commented on index.js:
```diff
@@ -20,9 +20,23 @@ function codeImport(options = {}) {
+      const extractLines = (content, fromLine, toLine) => {
+        if (fromLine === undefined && toLine === undefined) { return content }
+        if (toLine === undefined) toLine = fromLine
+        const lines = content.split(/\r?\n/)
```
Maybe we could split it with `os.EOL`?

kevin940726 commented on index.js:
```diff
@@ -20,9 +20,23 @@ function codeImport(options = {}) {
       }
 
-      const filePath = fileMeta.slice('file='.length);
+      const res = /^file=(?<path>[^#]+)(?:#L(?<from>\d+))?(?::L(?<to>\d+))?$/.exec(fileMeta)
```
This will fail to match if the path itself contains `#` somewhere. Could we adjust this so that only the last `#` is accounted?

ostollmann replied:
Good catch. Fixed.

Note I also changed the test file to `say-#-hi.js` to test for this.

kevin940726 reviewed (changes requested):
Thank you for creating this PR!

What do you think if we also support `#L3:` pattern to slice the file from line 3 to the very end? I think it could be a very common use case when all the user wants is to strip out the import statements.

Just being curious, where do you know of this pattern `#Ls:Ln`? I feel like we can try to follow what GitHub does and use `-` to replace `:`. `#L3-` is also slightly more readable to me too.

ostollmann committed:
add node_modules to gitignore

ostollmann committed:
Change to dashes, make lookup lazy for #s in filepath, add preceding and following

ostollmann committed:
Add # in filename as test

ostollmann committed:
Add lines and line ranges to readme

ostollmann commented:
> Thank you for creating this PR!
> 

Np. I want to use it :)

> What do you think if we also support `#L3:` pattern to slice the file from line 3 to the very end? I think it could be a very common use case when all the user wants is to strip out the import statements.

That's a good idea. I added the ability to do both:
- preceding lines: `js file=./__fixtures__/say-#-hi.js#-L2`
- following lines: `js file=./__fixtures__/say-#-hi.js#L2-`

> 
> Just being curious, where do you know of this pattern `#Ls:Ln`? I feel like we can try to follow what GitHub does and use `-` to replace `:`. `#L3-` is also slightly more readable to me too.

I wanted to do what Github does and thought they used `?:?`. Should have checked. Changed to `?-?` now.

ostollmann commented:
I also added this to the readme.

kevin940726 commented:
I just want to let you know that I've been a little bit busy lately, and I'll get back to it as soon as I have some time. Thank you for all the good work and your patience :)

kevin940726 commented on README.md:
````diff
@@ -48,6 +48,24 @@ console.log('Hello remark-code-import!');
+```js file=./say-hi.js#L3-
+```
+
+```js file=./say-hi.js#-L6
````
I don't think this is useful as it's the same as `#L1-L6` and it complicates the rule a little bit. I believe we should remove it, WDYT?

ostollmann replied:
Indeed. I removed this and added a test case to check for it: https://github.com/o1iver/remark-code-import/blob/0051e274e9340b0be4b9db5e78641bc243e165d7/test.js#L65.

kevin940726 commented on index.js:
```diff
@@ -20,9 +21,24 @@ function codeImport(options = {}) {
+      const toLine = res.groups.to ? parseInt(res.groups.to) : undefined
       const fileAbsPath = path.resolve(file.dirname, filePath);
 
+      const extractLines = (content, fromLine, toLine) => {
```
Prefer to move this function outside the `codeImport` function and make it pure (side-effect-free) so that it's easier to reason about.

ostollmann replied:
Not quite sure what you mean with pure or what side-effect it was creating; it was always pure, even it defined inline (except for `hasDash` which it wasn't modifying). The args shadow the local vars so it's not even a closure. I extracted `extractLines` to the top level though if that's what you meant.

kevin940726 replied:
Yeah, I just prefer it to be outside the main function so that it's a little bit more clear for me to review 😅.

kevin940726 reviewed (changes requested):
Good work!

There are some styling issues and tooling issues, probably has something to do with yarn berry not working correctly on your machine. I can handle that after we merge this though.

ostollmann committed:
Remove ability to "-L10" and move extract function to top-level

kevin940726 reviewed (approved):
Thank you! There are some minor styling preference of mine. I'll update them. 👍
